lib/ansible/plugins/inventory/libvirt.py

```diff
@@ -0,0 +1,124 @@
+
+        valid = False
+        if super(InventoryModule, self).verify_file(path):
+            if path.endswith(('.libvirt.yaml', '.libvirt.yml')):
+                valid = True
+        return valid
+
+    def parse(self, inventory, loader, path, cache=True):
+        ''' Construct the inventory '''
+
+        if not HAS_LIBVIRT:
+            raise AnsibleParserError("This plugin requires 'libvirt' to operate")
+
+        super(InventoryModule, self).parse(inventory, loader, path)
+
+        self._config_data = self._read_config_data(path)
+        self.set_options(direct=self._config_data)
+
+        states = self.get_option('states')
+        uri = self.get_option('uri')
+
+        if not states:
+            raise AnsibleParserError("At least one state must be specified")
+        else:
+            invalid = set(states).difference(STATES)
+            if invalid:
+                raise AnsibleParserError("Invalid states supplied: %s" % invalid)
+
+        conn = libvirt.openReadOnly(uri)
+        if conn is None:
+            raise AnsibleParserError("Failed to open connection to %s" % uri)
+        domains = conn.listAllDomains()
+        if domains is None:
+            raise AnsibleParserError("Failed to list domains for connection %s" % uri)
+
+        for domain in domains:
+            host = domain.name()
+            state = domain.state()[0]
+            if STATES[state] not in states:
+                continue
+
+            # create host and add some vars
+            self.inventory.add_host(host)
+            self.inventory.set_variable(host, 'libvirt_status', STATES[state])
+            self.inventory.set_variable(host, 'libvirt_name', host)
+            self.inventory.set_variable(host, 'libvirt_id', domain.ID())
+            self.inventory.set_variable(host, 'libvirt_uuid', domain.UUIDString())
+
+            # parse xml for tags
+            root = ET.fromstring(domain.XMLDesc())
+            tags = []
+            for tag_elem in root.findall('./metadata/ansible:tags/ansible:tag', ansible_ns):
+                tags.append(tag_elem.text)
+            self.inventory.set_variable(host, 'libvirt_tags', tags)
+
+            # TODO: support more than one network interface, also support interface types other than 'network'
+            interface = root.find("./devices/interface[@type='network']")
+            if interface is not None:
+                source_elem = interface.find('source')
+                mac_elem = interface.find('mac')
+                if source_elem is not None and mac_elem is not None:
+                    dhcp_leases = conn.networkLookupByName(source_elem.get('network')).DHCPLeases(mac_elem.get('address'))
+                    if len(dhcp_leases) > 0:
+                        ip_address = dhcp_leases[0]['ipaddr']
+                        self.inventory.set_variable(host, 'ansible_host', ip_address)
+                        self.inventory.set_variable(host, 'libvirt_ip_address', ip_address)
```

Comment:
This will only get the ip address for non fixed ip's?

Another way of getting ip's from libvirt is

```python
domain.interfaceAddresses(libvirt.VIR_DOMAIN_INTERFACE_ADDRESSES_SRC_AGENT, 0)
{'lo': {'hwaddr': '00:00:00:00:00:00', 'addrs': [{'prefix': 8, 'type': 0, 'addr': '127.0.0.1'}, {'prefix': 128, 'type': 1, 'addr': '::1'}]}, 'eth0': {'hwaddr': '52:54:00:03:b1:0b', 'addrs': [{'prefix': 24, 'type': 0, 'addr': '192.168.122.2'}, {'prefix': 64, 'type': 1, 'addr': 'fe80::c2de:d88c:1c2f:21c5'}]}}
```

Reply:
it assumes the dhcp lease is the 'preffered IP' to interface with it, it should really be configurable but im not sure how to approach that